src/native/marker-index.cc

```diff
@@ -246,18 +262,26 @@ unordered_map<MarkerId, Range> MarkerIndex::Dump() {
 }
 
 Point MarkerIndex::GetNodePosition(const Node *node) const {
-  Point position = node->left_extent;
-  while (node->parent) {
-    if (node->parent->right == node) {
-      position = node->parent->left_extent.Traverse(position);
-    }
+  auto cache_entry = node_position_cache.find(node);
+  if (cache_entry == node_position_cache.end()) {
+    Point position = node->left_extent;
+    const Node *current_node = node;
+    while (current_node->parent) {
+      if (current_node->parent->right == current_node) {
+        position = current_node->parent->left_extent.Traverse(position);
+      }
 
-    node = node->parent;
+      current_node = current_node->parent;
+    }
+    node_position_cache.insert({node, position});
+    return position;
```

Comment:
:+1: Thanks for catching that. I think I started writing this thinking we could return the same variable at the end.